Separate instance name from node name

akshaymankar opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://git.k8s.io/community/contributors/devel/pull-requests.md#the-pr-submit-process and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/devel/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/devel/pull-requests.md#write-release-notes-if-needed
4. If the PR is unfinished, see how to mark it: https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md#marking-unfinished-pull-requests
-->

**What this PR does / why we need it**:

This change allows the node's instance name to be provided through the cloud config file when using the GCE cloud provider. That allows the node's hostname to be different from GCE instance name, which helps an operator to bring their own TLS certs for nodes.

**Special notes for your reviewer**:

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
This change is fully backwards compatible, if an operator wants to keep using the GCP instance-name as hostname and node-name, no extra configuration is required. 

However, if the operator wants to give different hostnames and bring their own TLS, they'd need  to:

1. Add `instance-name` to the GCE cloud config
2. Start kubelet with --hostname-override set to whatever hostname they would like the kubelet to be addressed with
```

akshaymankar commented:
Fixes #58084

akshaymankar commented:
Hi @thockin @mtaufen,

As discussed in #58084, We've separated hostname and instance-name for GCE. Please take a look and let us know if you have any feedback.

akshaymankar commented:
/assign @bowei

mtaufen commented:
/ok-to-test

mtaufen commented:
I think I like the idea of including the instance name in the cloud config file, since it's specific to the cloud provider anyway. @thockin wdyt?

akshaymankar commented:
/retest

akshaymankar commented:
/retest

akshaymankar commented:
/test pull-kubernetes-integration

akshaymankar commented:
/test pull-kubernetes-e2e-gce

Commit:
Add instanceName and instanceNameMap to GCECloud

This change allows the node's instance name to be provided through the
cloud config file. That allows the node's hostname to be
different from GCE instance name, which helps an operator to bring their
own TLS certs for nodes.

Due to the fact that the controller-manager cannot just assume node name
is the instance name, it must keep a mapping of node names to instance
names.

Signed-off-by: James Myers <jfmyers9@gmail.com>
Signed-off-by: James Myers <jmyers@pivotal.io>

Commit:
Add tests for adding instance name to GCECloud

Signed-off-by: Akshay Mankar <amankar@pivotal.io>
Signed-off-by: James Myers <jfmyers9@gmail.com>
Signed-off-by: James Myers <jmyers@pivotal.io>

Commit:
Add tests for mapNodeNameToInstanceName

Signed-off-by: Iain Sproat <isproat@pivotal.io>

iainsproat committed:
Add tests for updating instanceNameMap

Signed-off-by: Akshay Mankar <amankar@pivotal.io>

Commit:
Use existing function to split providerID

Signed-off-by: Iain Sproat <isproat@pivotal.io>

mtaufen commented:
This is tangentially related: https://github.com/kubernetes/kubernetes/issues/61410

liggitt commented:
cc @mikedanese

Is this making the kubelet the authority on which cloud provider instance it is running on? I expected any mapping like that to come from the cloud node controller side, not be based on a field reported by the kubelet.

mtaufen commented:
How does this PR play with the `provider-id` flag? https://github.com/kubernetes/kubernetes/pull/44258

akshaymankar commented:
Hi @liggitt,

I am not quite sure what you mean by "making the kubelet the authority on which cloud provider instance it is running on?". 

As far as I understand kubelet takes these two flags`--cloud-provider` and `--cloud-config` to decide which cloud provider to use. This makes sure that kubelet reports with the right `spec.providerID` and then the controllers have their own cloud config provided to the kube-controller-manager. With this understanding, all we've done is to remove the assumption that the node name of the kubelet is the instance name of the VM it is running on.

I am not so sure about cloud node controller, can you please point me to a resource/code where I can learn more about it?

akshaymankar commented:
@mtaufen We tested the build from this PR with `--provider-id` and the cloud-controller-manager. We found that it doesn't work if you set `--hostname-override` to anything other than the instance name of the node VM. This is because of the same assumption that node names are instance names at these places:

* [gce_loadbalancer_external.go#L52-L54](https://github.com/kubernetes/kubernetes/blob/v1.10.0/pkg/cloudprovider/providers/gce/gce_loadbalancer_external.go#L52-L54)
* [gce_loadbalancer_external.go#L279](https://github.com/kubernetes/kubernetes/blob/v1.10.0/pkg/cloudprovider/providers/gce/gce_loadbalancer_external.go#L279)
* [gce_instances.go#L630](https://github.com/kubernetes/kubernetes/blob/v1.10.0/pkg/cloudprovider/providers/gce/gce_instances.go#L630)

We will work to change at these places and try to figure out if we have missed any other places.

Thanks for the heads up!

Commit:
Manage LBs by instance names instead of node names

For this to work, we needed to set informers on cloud provider in cloud
controller manager. This makes sure that the instanceNameMap in
GCECloud is up to date when the controllers want to know instance name
of a node.

Signed-off-by: Urvashi Reddy <ureddy@pivotal.io>

akshaymankar commented:
We've fixed the three places mentioned above. The loadbalancer got created by the cloud-controller-manager. We will do some more exploring around cloud-controller-manager to make sure we haven't missed any places where nodename=instancename assumption exists.

To make it work in cloud-controller-manger, we had to call `SetInformers` on the cloud object, not sure if that was [the right place](https://github.com/kubernetes/kubernetes/pull/61042/files#diff-43975f373e683caa7a5aad92f5a7c4f1R209) to do it. @mtaufen can you please provide any suggestions on this?

akshaymankar commented:
/assign @wlan0

liggitt commented on pkg/cloudprovider/providers/gce/gce.go:
```diff
@@ -724,6 +746,30 @@ func (gce *GCECloud) updateNodeZones(prevNode, newNode *v1.Node) {
+	defer gce.instanceNameMapLock.Unlock()
+
+	if newNode != nil {
+		_, _, instanceName, err := splitProviderID(newNode.Spec.ProviderID)
```
this is changing the cloud controllers to *pull* information from the ProviderID field of a Node API object, correct? doesn't that make the kubelet the authority on which cloud provider instance a given Node API object corresponds to?
@mikedanese ^

akshaymankar replied:
The only thing consumed from ProviderID field here is the instance name, which was assumed to be the node name before this by the cloud controllers. So, the kubelet was and still is the the authority in defining the node name, which was assumed to be the instance name by the cloud controllers.

liggitt replied:
The kubelet does not get to choose any node metadata.name it wants. It is only allowed to modify the one that corresponds to its kube API credentials. 

If this is changing cloudprovider code to be driven by the provider id field instead, we first have to ensure a compromised kubelet cannot claim to be another instance (by setting the providerID field) and get the cloud controller (or other controllers) to take actions against that instance

akshaymankar replied:
Hmm, I get it now. I was always assuming that kubelets were trusted as far as they were authenticated. I am not sure how to solve this problem, I will try to dig into other cloud providers to see how they solve it. Meanwhile, if you have any suggestions, please let me know.

akshaymankar replied:
I checked in other cloud providers (AWS, Openstack, Azure and vSphere), it seems like only vSphere relies on ProviderID and hence is insecure against a node being compromised. 

After chatting with a few people an idea that seems to work is tagging nodes in IAAS to make sure kubelet is not telling cloud controllers which nodes it corresponds to. For example, `node-1` runs on instance `vm-foo`. In this case `vm-foo` will be expected to have a tag say `kubernetes-node-name: node-1`. When node-1 joins the cluster, the cloud provider will look for vm with the said tag and find the instance name, from this point on the cloud provider can keep a mapping of node name to instance name.

This seems to solve the trust issue between master components and kubelet, assuming the tags can be trusted. But it will require operator to make extra configurations. I am not sure if that is acceptable. WDYT?

akshaymankar replied:
With labels, the code for `addNodeToInstanceMap` will look like this:
```golang
	if newNode != nil {
		_, zone, instanceName, err := splitProviderID(newNode.Spec.ProviderID)
		if err != nil {
			return err
		}
		instance, err := gce.Compute().Instances().Get(context.TODO(), &meta.Key{Name: instanceName, Zone: zone})
		if err != nil {
			return err
		}
		allowedNodeName, ok := instance.Labels["kubernetes-node-name"]
		if !ok {
			return errors.Errorf("Could not find label: 'kubernetes-node-name' on instance '%s'", instanceName)
		}

		if allowedNodeName != newNode.Name {
			return errors.Errorf("Instance '%s' is only allowed to have node name '%s'", instanceName, allowedNodeName)
		}

		gce.instanceNameMap[allowedNodeName] = instanceName
	}
```

liggitt replied:
> an idea that seems to work is labelling nodes in IAAS to make sure kubelet is not telling cloud controllers which nodes it corresponds to

having data living in the cloud provider being the source of truth is much closer to what I expected.

with that information living there, is it actually still necessary to have the ProviderID in the Node object? it's tough to ensure that every component that makes use of that field does a verification against a central source of truth that the mapping is actually correct.

akshaymankar replied:
I can think of two reasons for ProviderID to be on the Node Object:
1. It helps the cloud provider look at just one node and verify if the information is correct.
1. If somebody wants to recreate the worker while previous one is not gone, they can smoothly switch over by creating the new worker and then killing the previous one.

>it's tough to ensure that every component that makes use of that field does a verification against a central source of truth that the mapping is actually correct

No component (except for the `addNodeToInstanceMap` method) should be using ProviderID directly, they should always use the `mapNodeNameToInstanceName` method to get the instance name and the method above will make sure that the mapping is always verified.

akshaymankar replied:
I agree, it is weird to have that implicit assumption. Maybe the two points I listed above could be handled in other ways. 

It seems to be a little tricky to list VMs, because the GCE API to list vms only lists VMs in a given zone. So, either we'll need to make the `Zones` field mandatory in the cloud config or look for the instance through all the zones of GCE, which sounds like an overkill.

mikedanese commented:
> This change allows the node's instance name to be provided through the cloud config file when using the GCE cloud provider. That allows the node's hostname to be different from GCE instance name, which helps an operator to bring their own TLS certs for nodes.

Can you give a full explanation of the problem you are facing?

akshaymankar commented:
@mikedanese 

The problem we are facing is to generate certificates for kubelets, so we can secure communication between apiserver and kubelet. We were planning to use `--hostname-override` on kubelets to set the hostnames for apiserver -> kubelet communications to node1, node2, node3, etc. This also has a  side effect that node name changes to host name of the node. But we also discovered that the GCE cloud provider assumes that node name is the instance name in GCE. So, the `--hostname-override` needs to be set to instance name. This makes it impossible to generate certificates for kubelet before spinning up the VMs.

So, after some discussions on slack, we decided that it made sense to break this assumption and so we created the issue and this PR.

akshaymankar commented:
/test pull-kubernetes-bazel-test
/test pull-kubernetes-typecheck
/test pull-kubernetes-integration

liggitt commented:
fyi, I'm documenting the current state of cloud provider node name computation and requirements in https://github.com/kubernetes/website/pull/8873

once this issue is resolved, please update the GCP section of that document to describe the new method, any configuration requirements, etc.

liggitt commented on pkg/cloudprovider/providers/gce/gce.go:
```diff
@@ -179,6 +184,8 @@ type ConfigGlobal struct {
 	// blank, then the local zone will be discovered via the metadata server.
 	LocalZone string `gcfg:"local-zone"`
+	// InstanceName specifies the GCE instance name of the node
+	InstanceName string `gcfg:"instance-name"`
```
is this the source of truth for the instance name? expected to be consumed by the kubelet and registered in the Node object for use by the cloud controller? is there any way for the controller to verify from information in the instance that Node X really does correspond to it?

akshaymankar replied:
Yes, this is the source of truth for instance name and is expected to be consumed by kubelet. I cannot think of any straightforward way to verify the information that node X really does correspond to it. Which is the reason why I was thinking that I should close this PR.

It does annoy me that node name cannot be different from IaaS identifiers as of now. But until we can think of anything else I guess operators will have to live with it.

akshaymankar commented:
/close
